### PR TITLE
ADD: handle --chown on URLs

### LINKED ISSUE #2729: ADD --chown should invalidate cached layers
**Description**

It is impossible to update the `--chown` on an `ADD line` when caching layers.

**Steps to reproduce the issue:**

```Dockerfile
FROM scratch

ADD --chown=1000:100 https://static.rust-lang.org/rustup/dist/x86_64-unknown-linux-gnu/rustup-init /tmp/rustup-init
```

```shell
buildah bud --layers -f Dockerfile
# change ADD line to --chown=1001:101
buildah bud --layers -f Dockerfile
```

**Describe the results you received:**

ADD instruction reuses cached layer the second time.

**Describe the results you expected:**

ADD instruction DOES NOT reuse cached layer the second time.

**Output of `rpm -q buildah` or `apt list buildah`:**

```
buildah-1.16.2-1.1.x86_64
```

**Output of `buildah version`:**

```
Version:         1.16.2
Go Version:      go1.14.9
Image Spec:      1.0.1-dev
Runtime Spec:    1.0.2-dev
CNI Spec:        0.4.0
libcni Version:  v0.7.2-0.20190904153231-83439463f784
image Version:   5.5.2
Git Commit:      unknown
Built:           Tue Sep 29 21:00:00 2020
OS/Arch:         linux/amd64
```

**Output of `cat /etc/*release`:**

```
NAME="openSUSE Tumbleweed"
# VERSION="20201025"
ID="opensuse-tumbleweed"
ID_LIKE="opensuse suse"
VERSION_ID="20201025"
PRETTY_NAME="openSUSE Tumbleweed"
ANSI_COLOR="0;32"
CPE_NAME="cpe:/o:opensuse:tumbleweed:20201025"
BUG_REPORT_URL="https://bugs.opensuse.org"
HOME_URL="https://www.opensuse.org/"
DOCUMENTATION_URL="https://en.opensuse.org/Portal:Tumbleweed"
LOGO="distributor-logo"
```

**Output of `uname -a`:**

```
Linux rivendell 5.8.15-1-default #1 SMP Thu Oct 15 08:10:08 UTC 2020 (c680e93) x86_64 x86_64 x86_64 GNU/Linux
```

**Output of `cat /etc/containers/storage.conf`:**

```
# This file is is the configuration file for all tools
# that use the containers/storage library.
# See man 5 containers-storage.conf for more information
# The "container storage" table contains all of the server options.
[storage]

# Default Storage Driver
driver = "btrfs"

# Temporary storage location
runroot = "/var/run/containers/storage"

# Primary Read/Write location of container storage
graphroot = "/var/lib/containers/storage"

[storage.options]
# Storage options to be passed to underlying storage drivers

# AdditionalImageStores is used to pass paths to additional Read/Only image stores
# Must be comma separated list.
additionalimagestores = [
]

# Size is used to set a maximum size of the container image.  Only supported by
# certain container storage drivers.
size = ""

# Path to an helper program to use for mounting the file system instead of mounting it
# directly.
#mount_program = "/usr/bin/fuse-overlayfs"

# OverrideKernelCheck tells the driver to ignore kernel checks based on kernel version
# override_kernel_check = "false"

# mountopt specifies comma separated list of extra mount options
# mountopt = "nodev"

# Remap-UIDs/GIDs is the mapping from UIDs/GIDs as they should appear inside of
# a container, to UIDs/GIDs as they should appear outside of the container, and
# the length of the range of UIDs/GIDs.  Additional mapped sets can be listed
# and will be heeded by libraries, but there are limits to the number of
# mappings which the kernel will allow when you later attempt to run a
# container.
#
# remap-uids = 0:1668442479:65536
# remap-gids = 0:1668442479:65536

# Remap-User/Group is a name which can be used to look up one or more UID/GID
# ranges in the /etc/subuid or /etc/subgid file.  Mappings are set up starting
# with an in-container ID of 0 and the a host-level ID taken from the lowest
# range that matches the specified name, and using the length of that range.
# Additional ranges are then assigned, using the ranges which specify the
# lowest host-level IDs first, to the lowest not-yet-mapped container-level ID,
# until all of the entries have been used for maps.
#
# remap-user = "storage"
# remap-group = "storage"

# If specified, use OSTree to deduplicate files with the overlay backend
ostree_repo = ""

# Set to skip a PRIVATE bind mount on the storage home directory.  Only supported by
# certain container storage drivers
# skip_mount_home = "false"

[storage.options.thinpool]
# Storage Options for thinpool

# autoextend_percent determines the amount by which pool needs to be
# grown. This is specified in terms of % of pool size. So a value of 20 means
# that when threshold is hit, pool will be grown by 20% of existing
# pool size.
# autoextend_percent = "20"

# autoextend_threshold determines the pool extension threshold in terms
# of percentage of pool size. For example, if threshold is 60, that means when
# pool is 60% full, threshold has been hit.
# autoextend_threshold = "80"

# basesize specifies the size to use when creating the base device, which
# limits the size of images and containers.
# basesize = "10G"

# blocksize specifies a custom blocksize to use for the thin pool.
# blocksize="64k"

# directlvm_device specifies a custom block storage device to use for the
# thin pool. Required if you setup devicemapper.
# directlvm_device = ""

# directlvm_device_force wipes device even if device already has a filesystem.
# directlvm_device_force = "True"

# fs specifies the filesystem type to use for the base device.
# fs="xfs"

# log_level sets the log level of devicemapper.
# 0: LogLevelSuppress 0 (Default)
# 2: LogLevelFatal
# 3: LogLevelErr
# 4: LogLevelWarn
# 5: LogLevelNotice
# 6: LogLevelInfo
# 7: LogLevelDebug
# log_level = "7"

# min_free_space specifies the min free space percent in a thin pool require for
# new device creation to succeed. Valid values are from 0% - 99%.
# Value 0% disables
# min_free_space = "10%"

# mkfsarg specifies extra mkfs arguments to be used when creating the base.
# device.
# mkfsarg = ""

# use_deferred_removal marks devicemapper block device for deferred removal.
# If the thinpool is in use when the driver attempts to remove it, the driver
# tells the kernel to remove it as soon as possible. Note this does not free
# up the disk space, use deferred deletion to fully remove the thinpool.
# use_deferred_removal = "True"

# use_deferred_deletion marks thinpool device for deferred deletion.
# If the device is busy when the driver attempts to delete it, the driver
# will attempt to delete device every 30 seconds until successful.
# If the program using the driver exits, the driver will continue attempting
# to cleanup the next time the driver is used. Deferred deletion permanently
# deletes the device and all data stored in device will be lost.
# use_deferred_deletion = "True"

# xfs_nospace_max_retries specifies the maximum number of retries XFS should
# attempt to complete IO when ENOSPC (no space) error is returned by
# underlying storage device.
# xfs_nospace_max_retries = "0"
```

### PR DESCRIPTION
**Pull request by nalind**
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

Incorporate the value of the --chown option when digesting content added from a URL, so that build caching will take it into account.  Add tests to ensure that layer caching takes into account the changed ownership of items that we're adding or copying.

#### How to verify it

New integration test!

#### Which issue(s) this PR fixes:

Fixes #2729.

#### Special notes for your reviewer:

Part of #2729 was fixed in #2659, but we weren't testing for it.

#### Does this PR introduce a user-facing change?

```
The build cache should take into account ownership differences due to ADD and COPY being used with the --chown flag.
```

### CODE REVIEW

**Comment by rhatdan** [comment]
LGTM
But I am not sure this fully fixes the issue.

**Comment by rhatdan** [comment]
Never mind, I have now proved to myself, that my tests were broken, it chown is working for the non url case.

**Comment by rhatdan** [comment]
@containers/podman-maintainers PTAL

**Comment by rhatdan** [comment]
@TomSweeneyRedHat PTAL

**Comment by nalind** [comment]
Rebased.

File: tests/bud.bats [hunk 1]
```diff
@@ -2312,3 +2312,69 @@ EOF
...
+  [[ "$status" -ne 0 ]]
+  run cmp prev1 prev2
+  [[ "$status" -ne 0 ]]
+  run cmp add1  add2
+  [[ "$status" -ne 0 ]]
+  cmp tar1 tar2
+  run cmp url1  url2
+  [[ "$status" -ne 0 ]]
+
+  # The first rounds of builds should all be different from each other, as a sanith thing.
```
Comment (TomSweeneyRedHat) [inline tests/bud.bats:2356]: nit of a nit
```suggestion
  # The first rounds of builds should all be different from each other, as a sanity thing.
```

**Comment by TomSweeneyRedHat** [comment]
LGTM
ignoring nit for the sake of bug fixing

**Pull request merged**